{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1e80d74d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "71903ce6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathToRfmixResults = \"/mnt/disks/local_ancestry_analysis/ukb_rfmix_results_matteo/\"\n",
    "outputPath = \"/mnt/disks/local_ancestry_analysis/ukb_tractor_analysis/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "68a7b87b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cholab    int64\n",
      "matteo    int64\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "# Take sample mapping and turn into list\n",
    "cho2Matteo = pd.read_csv(\"cholab_to_matteo_sampleMapping.txt\", delim_whitespace=True)\n",
    "cho2Matteo.head(5)\n",
    "print(cho2Matteo.dtypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "99711c00",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(488377, 2)\n",
      "(488172, 2)\n"
     ]
    }
   ],
   "source": [
    "print(cho2Matteo.shape)\n",
    "cho2Matteo = cho2Matteo[(cho2Matteo['matteo'] >= 0) & (cho2Matteo['cholab'] >= 0)]\n",
    "print(cho2Matteo.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7f3d21f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampleMap = dict(zip(cho2Matteo.matteo, cho2Matteo.cholab))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "171e0699",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to recode haplotype sample names from msp.tsv file\n",
    "def recodeSamples(sampleMap, admixed_samples_genopheno_matched, samplesList):\n",
    "    samples_cho = [] # Recoded samples\n",
    "    samples_rfmix_idxs_to_read = [] # Indexes of samples (ie columns) in rfmix results that have matching genotype data + phenotype\n",
    "    samples_cho_matched = [] \n",
    "    \n",
    "\n",
    "    for idx in range(len(samplesList)):\n",
    "        sample = samplesList[idx]\n",
    "        sample_matteo = int(sample.split('_')[0])\n",
    "        sample_cho = sampleMap.get(sample_matteo)\n",
    "        \n",
    "        if (sample_cho != None) and (sample_cho in admixed_samples_genopheno_matched):\n",
    "            sample_string = str(sample_cho) + \"_\" + str(sample_cho)\n",
    "            samples_cho.append(sample_string + \".0\")\n",
    "            samples_cho.append(sample_string + \".1\")\n",
    "            \n",
    "            samples_cho_matched.append(sample_cho)\n",
    "            \n",
    "            # Store index of column mapping to sample id to read\n",
    "            samples_rfmix_idxs_to_read.append(2 * idx + 6) # Offset \n",
    "            samples_rfmix_idxs_to_read.append(2 * idx + 6 + 1) \n",
    "            \n",
    "            \n",
    "    return samples_cho_matched, samples_cho, samples_rfmix_idxs_to_read"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1cf2cefe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2539\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[5432249, 5093858, 3329090, 3640465, 4911551]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in samples of admixed samples with no missing data (sample order as genotype files)\n",
    "fname = \"/mnt/disks/local_ancestry_analysis/ukb_tractor_analysis/input_data/ukb_hap_chr19_v2_admixed_final.psam\"\n",
    "admixed_samples_genopheno_matched =  pd.read_csv(fname, delim_whitespace=True).iloc[:, 0].values.tolist()\n",
    "print(len(admixed_samples_genopheno_matched))\n",
    "admixed_samples_genopheno_matched[:5] # Note: these are sample ids from our lab's sample files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4b3088e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9979 9978\n",
      "9999 9998\n"
     ]
    }
   ],
   "source": [
    "# Check column mismatch in msp files\n",
    "for binIndex in range(1,3):\n",
    "    fname = \"pca.chr19.bin\" + str(binIndex) + \".msp.tsv.gz\"\n",
    "    # Read in columns only\n",
    "    tmp1 = pd.read_csv(pathToRfmixResults + fname , delim_whitespace=True, skiprows=1, nrows=1)\n",
    "    tmp2 = pd.read_csv(pathToRfmixResults + fname , delim_whitespace=True, skiprows=2, nrows=1) # Correct columns\n",
    "    \n",
    "    print(len(tmp1.columns), len(tmp2.columns)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1176296a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create dictionary of datatypes for data in msp file (haplotype data needs to be integers only)\n",
    "def getDatatypes(sample_columns):\n",
    "    rfmixResults_datatypes = dict.fromkeys(sample_columns, np.int8)\n",
    "    rfmixResults_datatypes[0] = np.int8 #chm\n",
    "    rfmixResults_datatypes[1] = np.int64 #spos\n",
    "    rfmixResults_datatypes[2] = np.int64 #epos\n",
    "    rfmixResults_datatypes[3] = np.float64 #sgpos\n",
    "    rfmixResults_datatypes[4] = np.float64 #egpos\n",
    "    rfmixResults_datatypes[5] = np.int64 #nsnps\n",
    "    \n",
    "    return rfmixResults_datatypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "70bef4de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map individuals from RFMIX results (Matteo back to our sample index mapping)\n",
    "numberOfBins = 96 # partitions of UKB individuals\n",
    "rfmixResults_admixedAfricanOnly = []\n",
    "samples_cho_matched = []\n",
    "for binIndex in range(1, numberOfBins + 1): \n",
    "    fname = \"pca.chr19.bin\" + str(binIndex) + \".msp.tsv.gz\"\n",
    "    \n",
    "    # Read in columns only--> extra column in header that does not match with actual data columns\n",
    "    rfmixResults_columns = pd.read_csv(pathToRfmixResults + fname , delim_whitespace=True, skiprows=1, nrows=1).columns\n",
    "    rfmixResults_samples = rfmixResults_columns[7::2] # Get sample ids of individuals in this file\n",
    "\n",
    "    \n",
    "    # Rename samples to match genotype data and get columns of samples to get from msp file \n",
    "    samples_matched, samples_cho, samples_rfmix_matched_idxs = recodeSamples(sampleMap, admixed_samples_genopheno_matched, rfmixResults_samples)    \n",
    "    \n",
    "    rfmixResultsColumnToRead = [i for i in range(6)] + samples_rfmix_matched_idxs # Update column names and get samples to read\n",
    "\n",
    "    dataypes = getDatatypes(samples_rfmix_matched_idxs)\n",
    "    # Read in rfmix results (only matched samples) in full\n",
    "    rfmixResults = pd.read_csv(pathToRfmixResults + fname , delim_whitespace=True,\n",
    "                               skiprows=2, usecols=rfmixResultsColumnToRead, dtype=dataypes)\n",
    "    \n",
    "    # Rename sample ids (from matteo to cholab)\n",
    "    rfmixResults.columns = ['#chm', 'spos', 'epos', 'sgpos', 'egpos', 'nsnps'] + list(samples_cho)\n",
    "    \n",
    "    # Keep dataframe\n",
    "    rfmixResults_admixedAfricanOnly.append(rfmixResults)\n",
    "    samples_cho_matched += samples_matched\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6fc8fb35",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AFR + EUR:  26\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  27\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  22\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  29\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  34\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  24\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  30\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  33\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  19\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  28\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  29\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  35\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  31\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  23\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  25\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  23\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  23\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  38\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  21\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  25\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  28\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  24\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  20\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  33\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  23\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  31\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  29\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  22\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  23\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  25\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  16\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  26\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  28\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  32\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  24\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  25\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  22\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  24\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  23\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  30\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  27\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  35\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  22\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  28\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  32\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  26\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  37\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  36\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  26\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  25\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  33\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  38\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  31\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  17\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  31\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  19\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  28\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  26\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  27\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  16\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  15\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  24\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  23\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  25\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  19\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  26\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  30\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  21\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  19\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  17\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  27\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  18\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  27\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  25\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  26\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  23\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  21\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  32\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  28\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  27\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  31\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  34\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  22\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  24\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  24\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  33\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  23\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  25\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  35\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  32\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  26\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  24\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  27\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  25\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  32\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n",
      "AFR + EUR:  35\n",
      "AFR only:  0\n",
      "EUR only:  0\n",
      " w/AMR:  0\n",
      " w/SAS:  0\n",
      " w/EAS:  0\n"
     ]
    }
   ],
   "source": [
    "## Sanity check on rfmix results: only african and european admixed individuals should appear\n",
    "\n",
    "# Generate counts of haplotypes inherited from each ancestral population\n",
    "for i in range(len(rfmixResults_admixedAfricanOnly)):\n",
    "    \n",
    "    rfmixResults_tractorIn = rfmixResults_admixedAfricanOnly[i]\n",
    "    \n",
    "    individual_ancestry_counts = {}\n",
    "    individual_ancestry_groups = {}\n",
    "    \n",
    "    haplotypes = rfmixResults_tractorIn.columns[6:]\n",
    "    \n",
    "    # Check at ancestral breakdown for each individual\n",
    "    for haplotype in haplotypes:\n",
    "        sample_id = haplotype.split(\"_\")[0]\n",
    "        if sample_id in individual_ancestry_counts:\n",
    "            individual_ancestry_counts[sample_id] += rfmixResults_tractorIn[haplotype].value_counts()\n",
    "        else:\n",
    "            individual_ancestry_counts[sample_id] = rfmixResults_tractorIn[haplotype].value_counts()\n",
    "\n",
    "        if sample_id in individual_ancestry_groups:\n",
    "            groups = individual_ancestry_groups[sample_id]\n",
    "            individual_ancestry_groups[sample_id] =  groups.union(set(rfmixResults_tractorIn[haplotype].unique())) \n",
    "        else:\n",
    "            individual_ancestry_groups[sample_id] = set(rfmixResults_tractorIn[haplotype].unique())\n",
    "\n",
    "    eur_code = 3\n",
    "    afr_code = 0\n",
    "    amr_code = 1\n",
    "    sas_code = 4 # should not appear\n",
    "    eas_code = 2 # should not appear\n",
    "\n",
    "\n",
    "    samples_with_afr_eur_inherited_tracts = []\n",
    "    samples_with_afr_only_inherited_tracts = []\n",
    "    samples_with_eur_only_inherited_tracts = []\n",
    "    samples_with_amr_inherited_tracts = []\n",
    "    samples_with_sas_inherited_tracts = []\n",
    "    samples_with_eas_inherited_tracts = []\n",
    "\n",
    "\n",
    "    for sample_id, ancestry_groups in individual_ancestry_groups.items():\n",
    "        if set([0, 3]) == ancestry_groups:\n",
    "            samples_with_afr_eur_inherited_tracts.append(sample_id)\n",
    "\n",
    "        if set([0]) == ancestry_groups:\n",
    "            samples_with_afr_only_inherited_tracts.append(sample_id)\n",
    "\n",
    "        if set([3]) == ancestry_groups:\n",
    "            samples_with_eur_only_inherited_tracts.append(sample_id)\n",
    "\n",
    "    print(\"AFR + EUR: \", len(samples_with_afr_eur_inherited_tracts))\n",
    "    print(\"AFR only: \", len(samples_with_afr_only_inherited_tracts))\n",
    "    print(\"EUR only: \", len(samples_with_eur_only_inherited_tracts))\n",
    "    print(\" w/AMR: \", len(samples_with_amr_inherited_tracts))\n",
    "    print(\" w/SAS: \", len(samples_with_sas_inherited_tracts))\n",
    "    print(\" w/EAS: \", len(samples_with_eas_inherited_tracts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1a3f73e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pickle rfmixResults for reference\n",
    "with open(outputPath + 'rfmixResults_admixedAfricanOnly.pkl', 'wb') as f:\n",
    "    pickle.dump(rfmixResults_admixedAfricanOnly, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "73127956",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2539\n",
      "[4476280, 3845985, 3725323, 5004315, 4981593]\n"
     ]
    }
   ],
   "source": [
    "print(len(samples_cho_matched))\n",
    "print(samples_cho_matched[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a02e28af",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2539\n",
      "[5432249, 5093858, 3329090, 3640465, 4911551]\n",
      "2539\n",
      "[5432249, 5093858, 3329090, 3640465, 4911551]\n"
     ]
    }
   ],
   "source": [
    "# Get sample order in VCF files and remove individuals that do not rfmix results\n",
    "admixed_samples_genopheno_matched_rfmix = [sample for sample in admixed_samples_genopheno_matched if sample in samples_cho_matched]\n",
    "print(len(samples_cho_matched))\n",
    "print(admixed_samples_genopheno_matched[:5])\n",
    "print(len(admixed_samples_genopheno_matched_rfmix))\n",
    "print(admixed_samples_genopheno_matched_rfmix[:5]) # Use to reorder samples in filtered, and recoded rfmix results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f086a145",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output final list of individuals to text file for subsequent analysis extract (plink)\n",
    "fname_out = \"admixed_samples_chr19_nomiss_genotype_ordered_final.txt\"\n",
    "with open(outputPath + \"input_data/\" + fname_out, 'w') as f:\n",
    "    for sample in admixed_samples_genopheno_matched_rfmix:\n",
    "        sample_string = str(sample) + \" \" + str(sample)\n",
    "        f.write(sample_string + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "651a682b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(20228, 5084)\n",
      "Index(['#chm', 'spos', 'epos', 'sgpos', 'egpos', 'nsnps', '4476280_4476280.0',\n",
      "       '4476280_4476280.1', '3845985_3845985.0', '3845985_3845985.1',\n",
      "       ...\n",
      "       '2495681_2495681.0', '2495681_2495681.1', '5487635_5487635.0',\n",
      "       '5487635_5487635.1', '1948895_1948895.0', '1948895_1948895.1',\n",
      "       '2294175_2294175.0', '2294175_2294175.1', '4326419_4326419.0',\n",
      "       '4326419_4326419.1'],\n",
      "      dtype='object', length=5084)\n",
      "   #chm    spos     epos  sgpos  egpos  nsnps  4476280_4476280.0  \\\n",
      "0    19  601373   768721   1.59   2.46    100                0.0   \n",
      "1    19  768721   849618   2.46   3.08     45                0.0   \n",
      "2    19  849618  1090803   3.08   3.80    140                0.0   \n",
      "\n",
      "   4476280_4476280.1  3845985_3845985.0  3845985_3845985.1  \n",
      "0                3.0                0.0                3.0  \n",
      "1                3.0                0.0                3.0  \n",
      "2                3.0                0.0                3.0  \n"
     ]
    }
   ],
   "source": [
    "# Concatenate rfmix results of chromosome together\n",
    "rfmixResults_admixedAfricanOnly_allSamples = pd.concat(rfmixResults_admixedAfricanOnly)\n",
    "print(rfmixResults_admixedAfricanOnly_allSamples.shape)\n",
    "print(rfmixResults_admixedAfricanOnly_allSamples.columns)\n",
    "print(rfmixResults_admixedAfricanOnly_allSamples.iloc[:3, :10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "2b98e3ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[5432249, 5093858, 3329090, 3640465, 4911551]\n",
      "[5432249, 5093858, 3329090, 3640465, 4911551]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['5432249_5432249.0',\n",
       " '5432249_5432249.1',\n",
       " '5093858_5093858.0',\n",
       " '5093858_5093858.1',\n",
       " '3329090_3329090.0']"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reorder rfmix results based on order of samples in vcf file\n",
    "rfmixResults_admixedAfricanOnly_sample_ids = rfmixResults_admixedAfricanOnly_allSamples.columns[6::2]\n",
    "rfmixResults_admixedAfricanOnly_sample_ids = [int(id.split('_')[0]) for id in rfmixResults_admixedAfricanOnly_sample_ids]\n",
    "rfmixResults_admixedAfricanOnly_sample_ids[:5]\n",
    "\n",
    "reorder = {k:v for v,k in enumerate(admixed_samples_genopheno_matched_rfmix)}\n",
    "rfmixResults_admixedAfricanOnly_sample_ids.sort(key=reorder.get)\n",
    "\n",
    "print(admixed_samples_genopheno_matched_rfmix[:5])\n",
    "print(rfmixResults_admixedAfricanOnly_sample_ids[:5])\n",
    "\n",
    "# Convert sample ids back to rfmix format: sample_sample.0 and  sample_sample.1\n",
    "rfmixResults_admixedAfricanOnly_sample_ids_reorder = []\n",
    "for sample in rfmixResults_admixedAfricanOnly_sample_ids:\n",
    "    sample_string = str(sample) + \"_\" + str(sample)\n",
    "    rfmixResults_admixedAfricanOnly_sample_ids_reorder.append(sample_string + \".0\")\n",
    "    rfmixResults_admixedAfricanOnly_sample_ids_reorder.append(sample_string + \".1\")\n",
    "\n",
    "rfmixResults_admixedAfricanOnly_sample_ids_reorder[:5] # Use to reorder columns in rfmix results (concatenated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "f2358bb4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20228, 5084)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reorder columns of concatenated, filtered, and reordered rfmix results to order of geno\n",
    "reorderedColumns = list(rfmixResults_admixedAfricanOnly_allSamples.columns[0:6]) + list(rfmixResults_admixedAfricanOnly_sample_ids_reorder)\n",
    "\n",
    "rfmixResults_admixedAfricanOnly_allSamples_final = rfmixResults_admixedAfricanOnly_allSamples[reorderedColumns]\n",
    "\n",
    "rfmixResults_admixedAfricanOnly_allSamples_final.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "94a3d3f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   #chm    spos     epos  sgpos  egpos  nsnps  5432249_5432249.0\n",
      "0    19  601373   768721   1.59   2.46    100                NaN\n",
      "1    19  768721   849618   2.46   3.08     45                NaN\n",
      "2    19  849618  1090803   3.08   3.80    140                NaN\n"
     ]
    }
   ],
   "source": [
    "# Fill in NA values and recode ancestral codes (AFR = 0, EUR = 3 --> 1)\n",
    "print(rfmixResults_admixedAfricanOnly_allSamples_final.iloc[:3, :7])\n",
    "rfmixResults_admixedAfricanOnly_allSamples_final_int = rfmixResults_admixedAfricanOnly_allSamples_final.fillna(-1)\n",
    "# rfmixResults_admixedAfricanOnly_allSamples_final_int = rfmixResults_admixedAfricanOnly_allSamples_final_int.replace(3, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "5455737e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   #chm    spos     epos  sgpos  egpos  nsnps  5432249_5432249.0\n",
      "0    19  601373   768721   1.59   2.46    100                 -1\n",
      "1    19  768721   849618   2.46   3.08     45                 -1\n",
      "2    19  849618  1090803   3.08   3.80    140                 -1\n"
     ]
    }
   ],
   "source": [
    "# Change datypes to match output file format for msp data\n",
    "rfmixResults_datatypes = dict.fromkeys(rfmixResults_admixedAfricanOnly_allSamples_final_int.columns[6:], np.int8)\n",
    "rfmixResults_datatypes['#chm'] = np.int8 #chm\n",
    "rfmixResults_datatypes['spos'] = np.int64 #spos\n",
    "rfmixResults_datatypes['epos'] = np.int64 #epos\n",
    "rfmixResults_datatypes['sgpos'] = np.float64 #sgpos\n",
    "rfmixResults_datatypes['egpos'] = np.float64 #egpos\n",
    "rfmixResults_datatypes['nsnps'] = np.int64 #nsnps\n",
    "\n",
    "rfmixResults_admixedAfricanOnly_allSamples_final_int = rfmixResults_admixedAfricanOnly_allSamples_final_int.astype(rfmixResults_datatypes)\n",
    "print(rfmixResults_admixedAfricanOnly_allSamples_final_int.iloc[:3, :7])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "80eb964d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20228, 5084)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rfmixResults_admixedAfricanOnly_allSamples_final_int.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "d140d09c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chm</th>\n",
       "      <th>spos</th>\n",
       "      <th>epos</th>\n",
       "      <th>sgpos</th>\n",
       "      <th>egpos</th>\n",
       "      <th>nsnps</th>\n",
       "      <th>5432249_5432249.0</th>\n",
       "      <th>5432249_5432249.1</th>\n",
       "      <th>5093858_5093858.0</th>\n",
       "      <th>5093858_5093858.1</th>\n",
       "      <th>...</th>\n",
       "      <th>4563247_4563247.0</th>\n",
       "      <th>4563247_4563247.1</th>\n",
       "      <th>4145550_4145550.0</th>\n",
       "      <th>4145550_4145550.1</th>\n",
       "      <th>1816775_1816775.0</th>\n",
       "      <th>1816775_1816775.1</th>\n",
       "      <th>1571060_1571060.0</th>\n",
       "      <th>1571060_1571060.1</th>\n",
       "      <th>1599017_1599017.0</th>\n",
       "      <th>1599017_1599017.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>601373</td>\n",
       "      <td>768721</td>\n",
       "      <td>1.59</td>\n",
       "      <td>2.46</td>\n",
       "      <td>100</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>768721</td>\n",
       "      <td>849618</td>\n",
       "      <td>2.46</td>\n",
       "      <td>3.08</td>\n",
       "      <td>45</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19</td>\n",
       "      <td>849618</td>\n",
       "      <td>1090803</td>\n",
       "      <td>3.08</td>\n",
       "      <td>3.80</td>\n",
       "      <td>140</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>1090803</td>\n",
       "      <td>1252451</td>\n",
       "      <td>3.80</td>\n",
       "      <td>4.39</td>\n",
       "      <td>50</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>1252451</td>\n",
       "      <td>1343212</td>\n",
       "      <td>4.39</td>\n",
       "      <td>4.94</td>\n",
       "      <td>25</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>19</td>\n",
       "      <td>57527351</td>\n",
       "      <td>57674644</td>\n",
       "      <td>105.85</td>\n",
       "      <td>106.20</td>\n",
       "      <td>55</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>19</td>\n",
       "      <td>57674644</td>\n",
       "      <td>57784866</td>\n",
       "      <td>106.20</td>\n",
       "      <td>106.36</td>\n",
       "      <td>40</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>19</td>\n",
       "      <td>57784866</td>\n",
       "      <td>57946540</td>\n",
       "      <td>106.36</td>\n",
       "      <td>106.57</td>\n",
       "      <td>55</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>19</td>\n",
       "      <td>57946540</td>\n",
       "      <td>58537665</td>\n",
       "      <td>106.57</td>\n",
       "      <td>107.07</td>\n",
       "      <td>160</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>19</td>\n",
       "      <td>58537665</td>\n",
       "      <td>59097933</td>\n",
       "      <td>107.07</td>\n",
       "      <td>107.73</td>\n",
       "      <td>178</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20228 rows × 5084 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     #chm      spos      epos   sgpos   egpos  nsnps  5432249_5432249.0  \\\n",
       "0      19    601373    768721    1.59    2.46    100                 -1   \n",
       "1      19    768721    849618    2.46    3.08     45                 -1   \n",
       "2      19    849618   1090803    3.08    3.80    140                 -1   \n",
       "3      19   1090803   1252451    3.80    4.39     50                 -1   \n",
       "4      19   1252451   1343212    4.39    4.94     25                 -1   \n",
       "..    ...       ...       ...     ...     ...    ...                ...   \n",
       "213    19  57527351  57674644  105.85  106.20     55                 -1   \n",
       "214    19  57674644  57784866  106.20  106.36     40                 -1   \n",
       "215    19  57784866  57946540  106.36  106.57     55                 -1   \n",
       "216    19  57946540  58537665  106.57  107.07    160                 -1   \n",
       "217    19  58537665  59097933  107.07  107.73    178                 -1   \n",
       "\n",
       "     5432249_5432249.1  5093858_5093858.0  5093858_5093858.1  ...  \\\n",
       "0                   -1                 -1                 -1  ...   \n",
       "1                   -1                 -1                 -1  ...   \n",
       "2                   -1                 -1                 -1  ...   \n",
       "3                   -1                 -1                 -1  ...   \n",
       "4                   -1                 -1                 -1  ...   \n",
       "..                 ...                ...                ...  ...   \n",
       "213                 -1                 -1                 -1  ...   \n",
       "214                 -1                 -1                 -1  ...   \n",
       "215                 -1                 -1                 -1  ...   \n",
       "216                 -1                 -1                 -1  ...   \n",
       "217                 -1                 -1                 -1  ...   \n",
       "\n",
       "     4563247_4563247.0  4563247_4563247.1  4145550_4145550.0  \\\n",
       "0                   -1                 -1                 -1   \n",
       "1                   -1                 -1                 -1   \n",
       "2                   -1                 -1                 -1   \n",
       "3                   -1                 -1                 -1   \n",
       "4                   -1                 -1                 -1   \n",
       "..                 ...                ...                ...   \n",
       "213                 -1                 -1                 -1   \n",
       "214                 -1                 -1                 -1   \n",
       "215                 -1                 -1                 -1   \n",
       "216                 -1                 -1                 -1   \n",
       "217                 -1                 -1                 -1   \n",
       "\n",
       "     4145550_4145550.1  1816775_1816775.0  1816775_1816775.1  \\\n",
       "0                   -1                 -1                 -1   \n",
       "1                   -1                 -1                 -1   \n",
       "2                   -1                 -1                 -1   \n",
       "3                   -1                 -1                 -1   \n",
       "4                   -1                 -1                 -1   \n",
       "..                 ...                ...                ...   \n",
       "213                 -1                 -1                 -1   \n",
       "214                 -1                 -1                 -1   \n",
       "215                 -1                 -1                 -1   \n",
       "216                 -1                 -1                 -1   \n",
       "217                 -1                 -1                 -1   \n",
       "\n",
       "     1571060_1571060.0  1571060_1571060.1  1599017_1599017.0  \\\n",
       "0                   -1                 -1                 -1   \n",
       "1                   -1                 -1                 -1   \n",
       "2                   -1                 -1                 -1   \n",
       "3                   -1                 -1                 -1   \n",
       "4                   -1                 -1                 -1   \n",
       "..                 ...                ...                ...   \n",
       "213                 -1                 -1                 -1   \n",
       "214                 -1                 -1                 -1   \n",
       "215                 -1                 -1                 -1   \n",
       "216                 -1                 -1                 -1   \n",
       "217                 -1                 -1                 -1   \n",
       "\n",
       "     1599017_1599017.1  \n",
       "0                   -1  \n",
       "1                   -1  \n",
       "2                   -1  \n",
       "3                   -1  \n",
       "4                   -1  \n",
       "..                 ...  \n",
       "213                 -1  \n",
       "214                 -1  \n",
       "215                 -1  \n",
       "216                 -1  \n",
       "217                 -1  \n",
       "\n",
       "[20228 rows x 5084 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rfmixResults_admixedAfricanOnly_allSamples_final_int"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "3a80fd9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "fname_out = outputPath + \"input_data/\" \"pca.chr19.ukb_hap_admixed.mine.msp.tsv\"\n",
    "# Write supplemental information to file\n",
    "with open(fname_out, 'w') as f:\n",
    "    f.write('#Subpopulation order/codes: AFR=0       EUR=1\\n')\n",
    "# Write new file out\n",
    "rfmixResults_admixedAfricanOnly_allSamples_final_int.to_csv(fname_out, sep = \"\\t\", mode='a', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "d87e5be1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(20227, 5084)\n",
      "2539\n",
      "Index(['19', '601373', '768721', '1.59', '2.46', '100', '-1', '-1.1', '-1.2',\n",
      "       '-1.3',\n",
      "       ...\n",
      "       '-1.5016', '-1.5017', '-1.5018', '-1.5019', '-1.5020', '-1.5021',\n",
      "       '-1.5022', '-1.5023', '-1.5024', '-1.5025'],\n",
      "      dtype='object', length=5084)\n"
     ]
    }
   ],
   "source": [
    "# Test reading\n",
    "test = pd.read_csv(fname_out, sep=\"\\t\",\n",
    "                               skiprows=1) \n",
    "\n",
    "print(test.shape)\n",
    "print(len(test.columns[7::2]))\n",
    "print(test.columns) # Should be 2359 (ie number of individuals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "9793302f",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>19</th>\n",
       "      <th>601373</th>\n",
       "      <th>768721</th>\n",
       "      <th>1.59</th>\n",
       "      <th>2.46</th>\n",
       "      <th>100</th>\n",
       "      <th>-1</th>\n",
       "      <th>-1.1</th>\n",
       "      <th>-1.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>768721</td>\n",
       "      <td>849618</td>\n",
       "      <td>2.46</td>\n",
       "      <td>3.08</td>\n",
       "      <td>45</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>849618</td>\n",
       "      <td>1090803</td>\n",
       "      <td>3.08</td>\n",
       "      <td>3.80</td>\n",
       "      <td>140</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19</td>\n",
       "      <td>1090803</td>\n",
       "      <td>1252451</td>\n",
       "      <td>3.80</td>\n",
       "      <td>4.39</td>\n",
       "      <td>50</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>1252451</td>\n",
       "      <td>1343212</td>\n",
       "      <td>4.39</td>\n",
       "      <td>4.94</td>\n",
       "      <td>25</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>1343212</td>\n",
       "      <td>1550872</td>\n",
       "      <td>4.94</td>\n",
       "      <td>5.65</td>\n",
       "      <td>90</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>19</td>\n",
       "      <td>1550872</td>\n",
       "      <td>1664092</td>\n",
       "      <td>5.65</td>\n",
       "      <td>5.85</td>\n",
       "      <td>25</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>19</td>\n",
       "      <td>1664092</td>\n",
       "      <td>1750447</td>\n",
       "      <td>5.85</td>\n",
       "      <td>6.41</td>\n",
       "      <td>40</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>19</td>\n",
       "      <td>1750447</td>\n",
       "      <td>2014037</td>\n",
       "      <td>6.41</td>\n",
       "      <td>7.24</td>\n",
       "      <td>100</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>19</td>\n",
       "      <td>2014037</td>\n",
       "      <td>2252990</td>\n",
       "      <td>7.24</td>\n",
       "      <td>7.76</td>\n",
       "      <td>80</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>19</td>\n",
       "      <td>2252990</td>\n",
       "      <td>2501255</td>\n",
       "      <td>7.76</td>\n",
       "      <td>8.30</td>\n",
       "      <td>100</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   19   601373   768721  1.59  2.46  100  -1  -1.1  -1.2\n",
       "0  19   768721   849618  2.46  3.08   45  -1    -1    -1\n",
       "1  19   849618  1090803  3.08  3.80  140  -1    -1    -1\n",
       "2  19  1090803  1252451  3.80  4.39   50  -1    -1    -1\n",
       "3  19  1252451  1343212  4.39  4.94   25  -1    -1    -1\n",
       "4  19  1343212  1550872  4.94  5.65   90  -1    -1    -1\n",
       "5  19  1550872  1664092  5.65  5.85   25  -1    -1    -1\n",
       "6  19  1664092  1750447  5.85  6.41   40  -1    -1    -1\n",
       "7  19  1750447  2014037  6.41  7.24  100  -1    -1    -1\n",
       "8  19  2014037  2252990  7.24  7.76   80  -1    -1    -1\n",
       "9  19  2252990  2501255  7.76  8.30  100  -1    -1    -1"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.iloc[:10, :9]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "333f6dad",
   "metadata": {},
   "outputs": [],
   "source": [
    "rfmixResults_tractorIn = rfmixResults_admixedAfricanOnly[i]\n",
    "    \n",
    "    individual_ancestry_counts = {}\n",
    "    individual_ancestry_groups = {}\n",
    "    \n",
    "    haplotypes = rfmixResults_tractorIn.columns[6:]\n",
    "    \n",
    "    # Check at ancestral breakdown for each individual\n",
    "    for haplotype in haplotypes:\n",
    "        sample_id = haplotype.split(\"_\")[0]\n",
    "        if sample_id in individual_ancestry_counts:\n",
    "            individual_ancestry_counts[sample_id] += rfmixResults_tractorIn[haplotype].value_counts()\n",
    "        else:\n",
    "            individual_ancestry_counts[sample_id] = rfmixResults_tractorIn[haplotype].value_counts()\n",
    "\n",
    "        if sample_id in individual_ancestry_groups:\n",
    "            groups = individual_ancestry_groups[sample_id]\n",
    "            individual_ancestry_groups[sample_id] =  groups.union(set(rfmixResults_tractorIn[haplotype].unique())) \n",
    "        else:\n",
    "            individual_ancestry_groups[sample_id] = set(rfmixResults_tractorIn[haplotype].unique())\n",
    "\n",
    "    eur_code = 3\n",
    "    afr_code = 0\n",
    "    amr_code = 1\n",
    "    sas_code = 4 # should not appear\n",
    "    eas_code = 2 # should not appear\n",
    "\n",
    "\n",
    "    samples_with_afr_eur_inherited_tracts = []\n",
    "    samples_with_afr_only_inherited_tracts = []\n",
    "    samples_with_eur_only_inherited_tracts = []\n",
    "    samples_with_amr_inherited_tracts = []\n",
    "    samples_with_sas_inherited_tracts = []\n",
    "    samples_with_eas_inherited_tracts = []\n",
    "\n",
    "\n",
    "    for sample_id, ancestry_groups in individual_ancestry_groups.items():\n",
    "        if set([0, 3]) == ancestry_groups:\n",
    "            samples_with_afr_eur_inherited_tracts.append(sample_id)\n",
    "\n",
    "        if set([0]) == ancestry_groups:\n",
    "            samples_with_afr_only_inherited_tracts.append(sample_id)\n",
    "\n",
    "        if set([3]) == ancestry_groups:\n",
    "            samples_with_eur_only_inherited_tracts.append(sample_id)\n",
    "\n",
    "    print(\"AFR + EUR: \", len(samples_with_afr_eur_inherited_tracts))\n",
    "    print(\"AFR only: \", len(samples_with_afr_only_inherited_tracts))\n",
    "    print(\"EUR only: \", len(samples_with_eur_only_inherited_tracts))\n",
    "    print(\" w/AMR: \", len(samples_with_amr_inherited_tracts))\n",
    "    print(\" w/SAS: \", len(samples_with_sas_inherited_tracts))\n",
    "    print(\" w/EAS: \", len(samples_with_eas_inherited_tracts))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
